nwastdlib/asyncio_cache.py

```diff
@@ -0,0 +1,134 @@
+    if not pickled_value or not cache_checksum:
+        return None
+
+    cache_checksum = cache_checksum.decode("utf-8")
+    if (checksum := get_hmac_checksum(secret, pickled_value)) != cache_checksum:
+        logger.error(
+            "Checksum for cache was wrong, someone tampered with the values!",
+            correct_checksum=str(cache_checksum),
+            recalculated_checksum=checksum,
+        )
+        return None
+
+    try:
+        data = deserialize(pickled_value)
+    except Exception as e:
+        logger.error(
+            "Cache deserialization failed, returning None, so cache will be overwritten with a new value", error=e
+        )
+        return None
+    return data
+
+
+def cached_result(
+    pool: AIORedis, prefix: str, secret: str, key_name: str | None = None, expiry_seconds: int = 120
+) -> Callable:
+    """Pass returned result objects from a function call into redis.
+
+    Returns a decorator function that will cache every result of a function to redis. This only works
+    for functions with string, int or UUID arguments and result objects that can be serialized by the
+    python pickle library.
+
+    Example:
+        Decorate a suitable function with all the default settings like this:::
+            @cached_result(pool, "your_app_name", "SECRET_KEY_FOR_HMAC_CHECKSUM")
+            def my_cached_function(uuid_arg, string_arg, int_kwarg=1):
+                return do_stuff(uuid_arg, string_arg, kwarg)
+
+        The first call is cached and reused for 120s. If the defaults are inadequate use:::
+            @cached_result(pool, "your_app_name", "SECRET_KEY_FOR_HMAC_CHECKSUM", "desired_cache_key_name", 120)
+            def my_other_function...
+
+    Args:
+        pool: A async redis cache pool.
+        secret: used to generate a checksum too ensure cache tampering is not possible.
+        prefix: Prefix for the cache keys generated.
+        key_name: When specified the total redis key_name will be "prefix:key_name".
+        expiry_seconds: expiration in seconds. Defaults to two minutes (120s).
+
+    Returns:
+        decorator function
+
+    """
+
+    def cache_decorator(func: Callable) -> Callable:
+        @wraps(func)
+        async def func_wrapper(*args: tuple[Any], **kwargs: dict[str, Any]) -> Callable:
+            python_major, python_minor = sys.version_info[:2]
+            if key_name:
+                cache_key = f"{prefix}:{python_major}.{python_minor}:{key_name}"
+            else:
+                # Auto generate a cache key name based on function_name and a hash of the arguments
+                # Note: this makes no attempt to handle non-hashable values like lists and sets or other complex objects
+                args_and_kwargs_string = (args, frozenset(kwargs.items()))
+                cache_key = f"{prefix}:{python_major}.{python_minor}:{func.__name__}{args_and_kwargs_string}"
+                logger.info("CACHE KEY", cache_key=cache_key)
```

Comment:
Statement can be removed?

And perhaps do `log = logger.bind(cache_key=cache_key)` on the line below and use `log` on the following logger calls

Reply:
Will remove. As it could log potentially private arguments...